WIP: drm/vc4: Pass an update_id to the fw to return on completion

6by9 opened the pull request:
This should allow KMS to return the events correctly for
page flip completions.

ISSUE SEEN: On quiting X there appears to be a mode change
that is timing out because we aren't returning a completion.
I'm wondering if this is due to kicking off the mode change
when the update is still active, resulting in the fw never
giving us the appropriate callback.
DRM has a 10sec timeout which clears the issue, but it isn't
clean at present.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
@popcornmix This is the kernel side patch. Ignore the commit text - I think this should all be good now. I'll amend it and repush.

popcornmix commented:
What's the behaviour if kernel is updated without firmware? And vice versa?

6by9 commented:
Should all be fine.

Old firmware/new kernel: the fw will return a count ID of 0, which FKMS will take as "I don't support the new ID", and have the same behaviour as before.

New firmware/old kernel: the structure size has changed, so the fw will insert the count ID as being 0. On vsync it'll set that 0 into the top bits of the SMI reg passed across. The kernel will then never look at those registers.

popcornmix commented:
I'm seeing a problem here. With this patch and corresponding firmware, after exiting kodi the system crashes (kernel panic and loss of ssh connection).
Reverting either kernel or firmware to stock and the issue is gone.

You don't have to play a video in kodi. Just launch it and shutdown (S key) is enough.
Uart not working but panic looks roughly like:
```
Kernel panic - not syncing: System is deadlocked on memory
Backtrace:
(panic) form pagefault_out_of_memory
(out of memory)
__alloc_pages_nodemask
copy_process
_do_fork
sys_clone
```
(can also provide a screenshot).

6by9 committed:
media: bcm2835-unicam: Always service interrupts

From when bringing up the driver, there was a check in the isr
to ignore interrupts (claiming them handled) should the driver
not be streaming.

The VPU now will not register a camera driver if it finds a
CSI2 node enabled in device tree, therefore this flawed check is
redundant.

https://github.com/raspberrypi/linux/issues/3602

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Closing as not current work
